File: backend/app/schemas/content.py
from datetime import datetime

from pydantic import BaseModel


class EraOut(BaseModel):
    id: int
    slug: str
    name: str
    description: str
    order_index: int
    approx_date_start: str
    approx_date_end: str
    color_hex: str
    cover_image_url: str | None
    story_count: int = 0

    model_config = {"from_attributes": True}


class CharacterBrief(BaseModel):
    id: str
    slug: str
    name: str
    role: str
    image_url: str | None

    model_config = {"from_attributes": True}


class StoryBrief(BaseModel):
    id: str
    slug: str
    title: str
    summary: str
    order_index: int
    reading_time_minutes: int
    era_id: int
    themes: list[str]

    model_config = {"from_attributes": True}


class StoryOut(BaseModel):
    id: str
    slug: str
    title: str
    summary: str
    content: str
    order_index: int
    bible_references: list[str]
    themes: list[str]
    reading_time_minutes: int
    era: EraOut
    characters: list[CharacterBrief] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CharacterOut(BaseModel):
    id: str
    slug: str
    name: str
    also_known_as: list[str]
    birth_approx: str | None
    death_approx: str | None
    biography: str
    role: str
    significance: str
    bible_references: list[str]
    image_url: str | None
    era: EraOut
    stories: list[StoryBrief] = []

    model_config = {"from_attributes": True}


class TimelineEventOut(BaseModel):
    id: str
    era_id: int
    story_id: str | None
    character_id: str | None
    title: str
    description: str
    event_type: str
    year_approx: int
    year_display: str
    duration_years: int | None
    order_index: int

    model_config = {"from_attributes": True}


class PaginatedStories(BaseModel):
    items: list[StoryBrief]
    total: int
    page: int
    page_size: int
    pages: int


class PaginatedCharacters(BaseModel):
    items: list[CharacterBrief]
    total: int
    page: int
    page_size: int
    pages: int

File: frontend/src/types/content.ts
export interface Era {
  id: number;
  slug: string;
  name: string;
  description: string;
  order_index: number;
  approx_date_start: string;
  approx_date_end: string;
  color_hex: string;
  cover_image_url: string | null;
  story_count: number;
}

export interface StoryBrief {
  id: string;
  slug: string;
  title: string;
  summary: string;
  order_index: number;
  reading_time_minutes: number;
  era_id: number;
  themes: string[];
}

export interface CharacterBrief {
  id: string;
  slug: string;
  name: string;
  role: string;
  image_url: string | null;
}

export interface Story {
  id: string;
  slug: string;
  title: string;
  summary: string;
  content: string;
  order_index: number;
  bible_references: string[];
  themes: string[];
  reading_time_minutes: number;
  era: Era;
  characters: CharacterBrief[];
  created_at: string;
  updated_at: string;
}

export interface Character {
  id: string;
  slug: string;
  name: string;
  also_known_as: string[];
  birth_approx: string | null;
  death_approx: string | null;
  biography: string;
  role: string;
  significance: string;
  bible_references: string[];
  image_url: string | null;
  era: Era;
  stories: StoryBrief[];
}

export interface TimelineEvent {
  id: string;
  era_id: number;
  story_id: string | null;
  character_id: string | null;
  title: string;
  description: string;
  event_type: "event" | "birth" | "death" | "miracle" | "war" | "covenant" | "prophecy";
  year_approx: number;
  year_display: string;
  duration_years: number | null;
  order_index: number;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}
